testng-maven-eclipse-plugin/src/org/testng/eclipse/maven/MavenTestNGLaunchConfigurationProvider.java

```diff
@@ -1,89 +1,294 @@
+import org.eclipse.m2e.core.project.IMavenProjectFacade;
+import org.eclipse.m2e.profiles.core.internal.IProfileManager;
+import org.eclipse.m2e.profiles.core.internal.MavenProfilesCoreActivator;
+import org.eclipse.m2e.profiles.core.internal.ProfileData;
+import org.eclipse.m2e.profiles.core.internal.ProfileState;
 import org.testng.eclipse.launch.ITestNGLaunchConfigurationProvider;
 import org.testng.eclipse.launch.LaunchConfigurationHelper;
-import org.w3c.dom.Document;
-import org.w3c.dom.Node;
-import org.w3c.dom.NodeList;
 
 public class MavenTestNGLaunchConfigurationProvider implements ITestNGLaunchConfigurationProvider {
 
+  public static final Pattern PATTERN = Pattern.compile("\\$\\{([^}]+)\\}");
+
   @Override
-  public String getVmArguments(ILaunchConfiguration configuration) throws CoreException {
-    String vmArgs = null;
-    try {
+  public String getVmArguments(ILaunchConfiguration launchConf) throws CoreException {
+    if (PreferenceUtils.getBoolean(Activator.PREF_ARGLINE)
+        || PreferenceUtils.getBoolean(Activator.PREF_SYSPROPERTIES)) {
+      String vmArgs = getVMArgsFromPom(launchConf);
+      return vmArgs;
+    }
+    return null;
+  }
+
+  @Override
+  public List<String> getEnvironment(ILaunchConfiguration launchConf) throws CoreException {
+    IJavaProject javaProject = LaunchConfigurationHelper.getProject(launchConf);
+    IMavenProjectFacade prjFecade = MavenPlugin.getMavenProjectRegistry().getProject(javaProject.getProject());
+    IProfileManager profileManager = MavenProfilesCoreActivator.getDefault().getProfileManager();
+
+    List<ProfileData> profiles = profileManager.getProfileDatas(prjFecade, new NullProgressMonitor());
+
+    Model model = MavenPlugin.getMavenModelManager().readMavenModel(prjFecade.getPom());
+    Xpp3Dom confDom = findPluginConfiguration(model, profiles);
+    if (confDom != null) {
+      if (PreferenceUtils.getBoolean(Activator.PREF_ENVIRON)) {
+        Xpp3Dom envDom = confDom.getChild("environmentVariables");
+        List<String> environList = new ArrayList<>(envDom.getChildCount());
+        for (Xpp3Dom varDom : envDom.getChildren()) {
+          environList.add(varDom.getName() + "=" + varDom.getValue());
+        }
+        return environList;
+      }
+    }
+    return null;
+  }
+
+  @SuppressWarnings("restriction")
+  private String getVMArgsFromPom(ILaunchConfiguration launchConf) throws CoreException {
+    IJavaProject javaProject = LaunchConfigurationHelper.getProject(launchConf);
+    IMavenProjectFacade prjFecade = MavenPlugin.getMavenProjectRegistry().getProject(javaProject.getProject());
+    IProfileManager profileManager = MavenProfilesCoreActivator.getDefault().getProfileManager();
+
+    List<ProfileData> selectedProfiles = profileManager.getProfileDatas(prjFecade, new NullProgressMonitor());
+
+    Model model = MavenPlugin.getMavenModelManager().readMavenModel(prjFecade.getPom());
+    Xpp3Dom confDom = findPluginConfiguration(model, selectedProfiles);
```

Comment:
find the plugin configuration for maven-surefire-plugin or maven-failsafe-plugin in the order of: 1) project pom.xml build base; 2) selected profiles (see m2e Maven Profile Management)